### PR TITLE
Introduce `Maybe` optional type

### PR DESCRIPTION
**Pull request by ebfull**
See #39.

The reason this came about is because we're implementing algorithms for field arithmetic in one of our crates and the current practice (as seen in `curve25519-dalek` for example) of

1. If you call `invert()` it returns zero for a zero input, when it should actually fail.
2. If you call inverse square root, it will return a value and a `Choice` indicating whether it's correct or not; the caller could accidentally forget.

is error prone. Given this new `Maybe` abstraction

1. Inversion can now return a `Maybe<T>` and if the caller is sure that the inputs are never zero, they can `unwrap()` themselves to make it explicit.
2. (Inverse) square root can return a `Maybe<T>` and the caller can choose how to deal with it. If they want to deal with the case in constant time, they can.

The API in this PR is really simple and could probably be extended to support e.g. constant time mapping over the underlying value, but I want to avoid blocking this feature on the inevitable questions of how such APIs should behave. What's implemented in this PR should be sufficient for all of the users of `subtle` I know about, and we can extend it later without breaking backwards compatibility.

The only question is whether or not the name `Maybe` is good.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -474,3 +474,68 @@ where
...
+
+impl<T: ConstantTimeEq> ConstantTimeEq for Maybe<T> {
+    /// Two `Maybe<T>`s are equal if they are both `Some` and
+    /// their values are equal, or both `None`.
+    #[inline]
+    fn ct_eq(&self, rhs: &Maybe<T>) -> Choice {
+        let a = self.is_some();
+        let b = rhs.is_some();
+
+        (self.is_some() & rhs.is_some() & self.value.ct_eq(&rhs.value)) | (!a & !b)
```
Comment (str4d) [inline src/lib.rs]: Why not `(a & b & self.value.ct_eq(&rhs.value)) | (!a & !b)`?

Comment (ebfull) [reply src/lib.rs]: I was in the middle of simplifying the function and got distracted! Thanks.

**Comment by daira** [comment]
I don't see how I would use this in constant time, given that I can't get the value when the choice is None. Constant-time operations usually require that you have a dummy value to operate on at the same time as propagating the choice to indicate that it's a dummy.

**Comment by ebfull** [comment]
I pushed to this PR implementations of `unwrap_or` and `unwrap_or_else` to demonstrate the intended behavior of this type. I just don't want to use this PR to implement constant time mapping because the API may require some debate and discussion.

**Comment by oleganza** [comment]
>@daira: I don't see how I would use this in constant time, given that I can't get the value when the choice is None. Constant-time operations usually require that you have a dummy value to operate on, at the same time as propagating the choice to indicate that it's a dummy.

@hdevalence suggestion was to place bound `T:Default` or similar to get such dummy value.

**Comment by oleganza** [comment]
We are implementing a logical `not` in ZkVM (https://github.com/interstellar/slingshot/issues/117) and we come to need this Maybe type for `Scalar::invert()`. 

For now, I guess the workaround is to have a conditional selection between `x` and `1`, and then invert unconditionally one of those.

**Comment by hdevalence** [comment]
I think some version of this is a good idea to have, but I have a few questions about details:

* Should the inner `T` value be bounded by `Default`? We can always remove a bound later, but if we add it without the bound, it's not possible to add the bound later without breaking stability. If we don't have a `Default` bound, it's not possible to construct a `None` without having the user supply a dummy value.

* Would it be better to call it `ConstantTimeOption<T>` (or something else) rather than `Maybe<T>`, so that it's immediately clear that it's a constant-time version of `Option` and has the same role?

* What subset of the `Option` methods make sense to provide constant-time analogs of?

**Comment by ebfull** [comment]
> Should the inner `T` value be bounded by `Default`?

I don't think it needs to be unless you're using a function, like `map` or `and_then`, which needs a dummy value to pass to the closure. We should place the bound on those functions specifically.

> Would it be better to call it `ConstantTimeOption<T>` (or something else) rather than `Maybe<T>`, so that it's immediately clear that it's a constant-time version of `Option` and has the same role?

That might be a better name. I wanted to avoid a long type name but it may be necessary. Perhaps there is a shorter alternative.

> What subset of the `Option` methods make sense to provide constant-time analogs of?

Definitely these

* `is_some` / `is_none` / `unwrap` (implemented already)
* `unwrap_or` / `unwrap_or_else` (implemented already)
* `expect`
* `map` / `map_or` / `map_or_else`
* `and` / `and_then`
* `or` / `or_else`
* `filter`
* `get_or_insert` / `get_or_insert_with` (I never knew about these)
* `take`
* `replace`

I'm not sure about these:

* `as_ref` / `as_mut`
* `ok_or` / `ok_or_else`
* `iter` / `iter_mut`

It should probably also implement `Clone` and all the other fun stuff.

**Comment by oleganza** [comment]
>That might be a better name. I wanted to avoid a long type name but it may be necessary. Perhaps there is a shorter alternative.

How about `ConstTimeOption` or `CtOption` in order to avoid three consecutive uppercase letters?

**Comment by ebfull** [comment]
I like `CtOption`.

**Comment by burdges** [comment]
Also `OptionCT` for "`Option` in constant-time".  It'd be `OptionTC` in French btw.  ;)

**Comment by oleganza** [comment]
lol.

**Comment by str4d** [comment]
Per the [Rust naming conventions](https://doc.rust-lang.org/1.0.0/style/style/naming/README.html):

> In `CamelCase`, acronyms count as one word: use `Uuid` rather than `UUID`.

So `CtOption` or `OptionCt` would fit. I prefer `CtOption` personally.

**Comment by hdevalence** [comment]
`CtOption` sounds great.

> > Should the inner `T` value be bounded by `Default`?
> 
> I don't think it needs to be unless you're using a function, like `map` or `and_then`, which needs a dummy value to pass to the closure. We should place the bound on those functions specifically.

Hmm, is it better to have a bound on those functions specifically (making the `CtOption` more general), or to have a unified set of functionality (making `CtOption` more restrictive)?  (I don't have a strong opinion either way).

> > What subset of the `Option` methods make sense to provide constant-time analogs of?
> 
> Definitely these (snipped)

This is a lot of functions, but they all seem sensible and it seems like a good idea to match the `Option` API. I can help implement some if you'd like.

> I'm not sure about these:
> 
>     * `as_ref` / `as_mut`

These might make sense, but I'm not really sure without thinking of particular applications, so maybe we could skip them, at least initially?

>     * `ok` / `ok_or` / `ok_or_else`

These convert to `Result`s, which aren't compatible with constant-time operations.  Maybe it would be better to have a single conversion from a `CtOption` into an `Option`, so that there's one way to cross out of the constant-time-world?

>     * `iter` / `iter_mut`

same considerations as above, but because iterators aren't constant time.

> It should probably also implement `Clone` and all the other fun stuff.

:+1:

**Comment by ebfull** [comment]
> Hmm, is it better to have a bound on those functions specifically (making the `CtOption` more general), or to have a unified set of functionality (making `CtOption` more restrictive)? (I don't have a strong opinion either way).

I would prefer to make the `CtOption` more general because in some cases I don't even need to use `map` (and similar functions) in any of my code, yet I'd still like to give the caller the opportunity to call e.g. `unwrap_or`. There seems to also be limited circumstances where the `Default` bound is needed: only functions like `map`.

**Comment by ebfull** [comment]
I pushed implementations of some additional things (including `map` and `and_then`) and changed the name to `CtOption`.

**Comment by hdevalence** [comment]
Added some commits to fix CI.

**Pull request merged**